### ISSUE TITLE
Hash operation consistency

### ISSUE THREAD

**Issue by bobbinth** [maticnetwork/miden#52]
There are currently some inconsistencies in how we've implemented hashing operations. Specifically:

* In the VM, `RPPERM` to compute *permute(A, B, C)*, expects the stack to be arranged as `[A, B, C, ...]` where *A* is at the top of the stack.
* In the assembly, `rpperm` and `rphash` operations assume that the stack is arranged as `[C, B, A, ...]` where *C* is the capacity portion of the sponge and is located at the top of the stack. In case of `rphash` we assume that once `RPPERM` is executed, the word *A* will contain the result of the hash.

One option to address this is to simply fix the implementation of `RPPERM` in the VM. However, it might be a good idea to have the capacity portion of the sponge deep in the stack (the way it is currently arranged for `RPPERM`). This way, when we compute a linear hash of many elements, we don't need to touch capacity elements (word *C*) as we update the outer portion of the sponge (words *A* and *B*).

At the same time, the arrangement `[A, B, C, ...]` is contrary to how we arrange stack for most other operations and might be confusing. For example, as we compute linear hash of many elements, we'd first need to populate the word *A* and only then the word *B* with new data - which will be awkward to do since `A` is at the top of the stack.

The other option is to modify how the Rp64_256 hash function actually works. Specifically, we can change it so that if the state is `[A, B, C]`, it is expected that `A` is the capacity portion, and `B` is expected to contain the result of the hash after the permutation.

So, to summarize the options:
1. Fix `RPPERM` operation in the VM. This is the simplest option but would make linear hashes of many elements slightly less efficient.
2. Update assembly instruction `rphash` to work with the current implementation of `RPPERM`. This accepts the awkwardness of the current stack arrangement of `RPPERM` and also would make computing linear hashes of many elements less efficient.
3. Update Rp64_256 hash function to work as described above, and update `RPPERM` and related assembly instructions accordingly. The stack would be arranged as: `[C, B, A, ...]`. This is a more involved option, but perhaps the cleanest and most efficient.

**Comment by grjte** [comment]
I guess it depends on how involved the Rp64_256 update would be, but I would lean towards preserving efficiency at this stage. It will certainly be easier to adjust now than later

**Comment by bobbinth** [comment]
@grjte Rp64_256 has been updated as a part of this PR: https://github.com/novifinancial/winterfell/pull/72

So, now we can update `winter-crypto` dependency to v0.3.2 and make the implementation of hashing consistent across various crates (basically, option 3 from above).

**Comment by grjte** [comment]
Great! Do you want me to take this on?

**Comment by bobbinth** [comment]
Yes! That would be great! I'll assign this issue to you.

**Comment by bobbinth** [comment]
Closed by #88
